components/f4jumble/src/lib.rs

```diff
@@ -40,92 +71,110 @@ macro_rules! G_PERS {
+    }
+
+    fn g_round(&mut self, i: u8) {
+        for j in 0..ceildiv(self.right.len(), OUTBYTES) {
+            let hash = Blake2bParams::new()
+                .hash_length(OUTBYTES)
+                .personal(&G_PERS!(i, j as u16))
+                .hash(&self.left);
+            xor(&mut self.right[j * OUTBYTES..], hash.as_bytes());
+        }
+    }
+
+    fn apply_f4jumble(&mut self) {
+        self.g_round(0);
+        self.h_round(0);
+        self.g_round(1);
+        self.h_round(1);
+    }
+
+    fn apply_f4jumble_inv(&mut self) {
+        self.h_round(1);
+        self.g_round(1);
+        self.h_round(0);
+        self.g_round(0);
     }
 }
 
-fn xor(a: &[u8], b: &[u8]) -> Vec<u8> {
-    a.iter().zip(b.iter()).map(|(a0, b0)| a0 ^ b0).collect()
+// xor bytes of the `source` to bytes of the `target`
+fn xor(target: &mut [u8], source: &[u8]) {
+    for (source, target) in source.iter().zip(target.iter_mut()) {
+        *target ^= source;
+    }
 }
 
 fn ceildiv(num: usize, den: usize) -> usize {
     (num + den - 1) / den
 }
 
-#[allow(clippy::many_single_char_names)]
-pub fn f4jumble(a: &[u8]) -> Option<Vec<u8>> {
-    if VALID_LENGTH.contains(&a.len()) {
-        let hashes = Hashes::new(a.len());
-        let (a, b) = a.split_at(hashes.l_l);
+pub fn f4jumble_mut(message: &mut [u8]) -> Result<(), Error> {
+    if VALID_LENGTH.contains(&message.len()) {
+        State::new(message).apply_f4jumble();
+        Ok(())
+    } else {
+        Err(Error::InvalidLength)
+    }
+}
 
-        let x = xor(b, &hashes.g(0, a));
-        let y = xor(a, &hashes.h(0, &x));
-        let d = xor(&x, &hashes.g(1, &y));
-        let mut c = xor(&y, &hashes.h(1, &d));
+pub fn f4jumble_inv_mut(message: &mut [u8]) -> Result<(), Error> {
+    if VALID_LENGTH.contains(&message.len()) {
+        State::new(message).apply_f4jumble_inv();
+        Ok(())
+    } else {
+        Err(Error::InvalidLength)
+    }
+}
 
-        c.extend(d);
-        Some(c)
+#[cfg(feature = "std")]
+pub fn f4jumble(message: &[u8]) -> Option<Vec<u8>> {
+    let mut result = message.to_vec();
+    let res = f4jumble_mut(&mut result);
+    if res.is_ok() {
+        Some(Vec::from(result))
     } else {
         None
     }
 }
 
-#[allow(clippy::many_single_char_names)]
-pub fn f4jumble_inv(c: &[u8]) -> Option<Vec<u8>> {
-    if VALID_LENGTH.contains(&c.len()) {
-        let hashes = Hashes::new(c.len());
-        let (c, d) = c.split_at(hashes.l_l);
-
-        let y = xor(c, &hashes.h(1, d));
-        let x = xor(d, &hashes.g(1, &y));
-        let mut a = xor(&y, &hashes.h(0, &x));
-        let b = xor(&x, &hashes.g(0, &a));
-
-        a.extend(b);
-        Some(a)
+#[cfg(feature = "std")]
+pub fn f4jumble_inv(message: &[u8]) -> Option<Vec<u8>> {
+    let mut result = message.to_vec();
+    let res = f4jumble_inv_mut(&mut result);
+    if res.is_ok() {
+        Some(Vec::from(result))
     } else {
         None
     }
 }
 
+#[cfg(feature = "std")]
```

Comment:
This means that the tests will not be run with no_std. So if there were a platform-dependent bug on an embedded platform, for example, we wouldn't catch it. Please add tests that run with no_std.